### ISSUE TITLE
Question of re-implement 

### ISSUE THREAD

**Issue by Youskrpig** [michuanhaohao/reid-strong-baseline#46]
Hi, thank you for the strong baseline.  
I reproduce the Experiment with tricks(Warmup + REA + LS + (Laststride = 1) + BNNECK) except the trick of Center Loss on Market1501, after 120 epoches the final results are as follow:
2019-06-25 16:03:10,081 reid_baseline.train INFO: mAP: 82.3%
2019-06-25 16:03:10,081 reid_baseline.train INFO: CMC curve, Rank-1  :92.5%

But as the paper said the mAP should be about 85%.  I'm so confused where am i going wrong？

**Comment by michuanhaohao** [comment]
How many iterations does an epoch have? We use new sample.py. Each epoch has 180+ iterations on Market1501.

**Comment by Simplesss** [comment]
Hello, I use the following parameters to train the model you provided, but the final result is very poor. It may be a problem with my parameter configuration. Can you provide your own training parameters?
2019-06-27 15:51:02,642 reid_baseline.train INFO: mAP: 66.7%
2019-06-27 15:51:02,642 reid_baseline.train INFO: CMC curve, Rank-1  :83.6%
Here is my defaults.py
from yacs.config import CfgNode as CN
# -----------------------------------------------------------------------------
# Config definition
# -----------------------------------------------------------------------------

_C = CN()

_C.MODEL = CN()
# Using cuda or cpu for training
_C.MODEL.DEVICE = "cuda"
# ID number of GPU
_C.MODEL.DEVICE_ID = '2'
# Name of backbone
_C.MODEL.NAME = 'resnet50'
# Last stride of backbone
_C.MODEL.LAST_STRIDE = 1
# Path to pretrained model of backbone
# _C.MODEL.PRETRAIN_PATH = '......./reid-strong-baseline/pretrained/market_resnet50_model_120_rank1_945.pth'
_C.MODEL.PRETRAIN_PATH = '...../reid-strong-baseline/checkpoint_mod/train/market1501_1/resnet50_model_120.pth'


# Use ImageNet pretrained model to initialize backbone or use self trained model to initialize the whole model
# Options: 'imagenet' or 'self' or 'others'
_C.MODEL.PRETRAIN_CHOICE = 'others'
# If train with BNNeck, options: 'bnneck' or 'no'
_C.MODEL.NECK = 'bnneck'
# If train loss include center loss, options: 'yes' or 'no'. Loss with center loss has different optimizer configuration
_C.MODEL.IF_WITH_CENTER = 'no'
# The loss type of metric loss
# options:'triplet','cluster','triplet_cluster','center','range_center','triplet_center','triplet_range_center'
_C.MODEL.METRIC_LOSS_TYPE = 'triplet_cluster'
# For example, if loss type is cross entropy loss + triplet loss + center loss
# the setting should be: _C.MODEL.METRIC_LOSS_TYPE = 'triplet_center' and _C.MODEL.IF_WITH_CENTER = 'yes'

# If train with label smooth, options: 'on', 'off'
_C.MODEL.IF_LABELSMOOTH = 'on'


# -----------------------------------------------------------------------------
# INPUT
# -----------------------------------------------------------------------------
_C.INPUT = CN()
# Size of the image during training
_C.INPUT.SIZE_TRAIN = [256, 128]
# Size of the image during test
_C.INPUT.SIZE_TEST = [256, 128]
# Random probability for image horizontal flip
_C.INPUT.PROB = 0.5
# Random probability for random erasing
_C.INPUT.RE_PROB = 0.5
# Values to be used for image normalization
_C.INPUT.PIXEL_MEAN = [0.485, 0.456, 0.406]
# Values to be used for image normalization
_C.INPUT.PIXEL_STD = [0.229, 0.224, 0.225]
# Value of padding size
_C.INPUT.PADDING = 10

# -----------------------------------------------------------------------------
# Dataset
# -----------------------------------------------------------------------------
_C.DATASETS = CN()
# List of the dataset names for training, as present in paths_catalog.py
_C.DATASETS.NAMES = ('market1501')
# Root directory where datasets should be used (and downloaded if not found)
_C.DATASETS.ROOT_DIR = ('....../reid-strong-basseline/data')

# -----------------------------------------------------------------------------
# DataLoader
# -----------------------------------------------------------------------------
_C.DATALOADER = CN()
# Number of data loading threads
_C.DATALOADER.NUM_WORKERS = 8
# Sampler for data loading
_C.DATALOADER.SAMPLER = 'softmax_triplet'
# Number of instance for one batch
_C.DATALOADER.NUM_INSTANCE = 4

# ---------------------------------------------------------------------------- #
# Solver
# ---------------------------------------------------------------------------- #
_C.SOLVER = CN()
# Name of optimizer
_C.SOLVER.OPTIMIZER_NAME = "Adam"
# Number of max epoches
_C.SOLVER.MAX_EPOCHS = 120
# Base learning rate
_C.SOLVER.BASE_LR = 3.5e-4
# Factor of learning bias
_C.SOLVER.BIAS_LR_FACTOR = 1
# Momentum
_C.SOLVER.MOMENTUM = 0.9
# Margin of triplet loss
_C.SOLVER.MARGIN = 0.3
# Margin of cluster ;pss
_C.SOLVER.CLUSTER_MARGIN = 0.3
# Learning rate of SGD to learn the centers of center loss
_C.SOLVER.CENTER_LR = 0.5
# Balanced weight of center loss
_C.SOLVER.CENTER_LOSS_WEIGHT = 0.0005
# Settings of range loss
_C.SOLVER.RANGE_K = 2
_C.SOLVER.RANGE_MARGIN = 0.3
_C.SOLVER.RANGE_ALPHA = 0
_C.SOLVER.RANGE_BETA = 1
_C.SOLVER.RANGE_LOSS_WEIGHT = 1

# Settings of weight decay
_C.SOLVER.WEIGHT_DECAY = 0.0005
_C.SOLVER.WEIGHT_DECAY_BIAS = 0.0005

# decay rate of learning rate
_C.SOLVER.GAMMA = 0.1
# decay step of learning rate
_C.SOLVER.STEPS = (40, 70)

# warm up factor
_C.SOLVER.WARMUP_FACTOR = 0.1
# iterations of warm up
_C.SOLVER.WARMUP_ITERS = 10
# method of warm up, option: 'constant','linear'
_C.SOLVER.WARMUP_METHOD = "linear"

# epoch number of saving checkpoints
_C.SOLVER.CHECKPOINT_PERIOD = 40
# iteration of display training log
_C.SOLVER.LOG_PERIOD = 20
# epoch number of validation
_C.SOLVER.EVAL_PERIOD = 40

# Number of images per batch
# This is global, so if we have 8 GPUs and IMS_PER_BATCH = 16, each GPU will
# see 2 images per batch
_C.SOLVER.IMS_PER_BATCH = 64

# This is global, so if we have 8 GPUs and IMS_PER_BATCH = 16, each GPU will
# see 2 images per batch
_C.TEST = CN()
# Number of images per batch during test
_C.TEST.IMS_PER_BATCH = 128
# If test with re-ranking, options: 'yes','no'
_C.TEST.RE_RANKING = 'no'
# Path to trained model
_C.TEST.WEIGHT = "......./reid-strong-baseline/checkpoint_mod/train/market1501_1/resnet50_model_120.pth"
# Which feature of BNNeck to be used for test, before or after BNNneck, options: 'before' or 'after'
_C.TEST.NECK_FEAT = 'after'
# Whether feature is nomalized before test, if yes, it is equivalent to cosine distance
_C.TEST.FEAT_NORM = 'yes'

# ---------------------------------------------------------------------------- #
# Misc options
# ---------------------------------------------------------------------------- #
# Path to checkpoint and saved log of trained model
_C.OUTPUT_DIR = "../checkpoint_mod/train/market1501_triplet_cluster"

I didn't load the .yml file, just all the parameters set in defaults.py, then run it in pycharm.
Thank you for your reply.

**Comment by Youskrpig** [comment]
Each Epoch has  203 Iterations.  Time per batch: 75.679[s]  Speed: 171.7[samples/s]

**Comment by Simplesss** [comment]
Hello, I use the following parameters to train the model you provided, but the final result is very poor. It may be a problem with my parameter configuration. Can you provide your own training parameters?
2019-06-27 15:51:02,642 reid_baseline.train INFO: mAP: 66.7%
2019-06-27 15:51:02,642 reid_baseline.train INFO: CMC curve, Rank-1  :83.6%
Here is my config
2019-06-27 11:40:13,190 reid_baseline INFO: Using 1 GPUS
2019-06-27 11:40:13,190 reid_baseline INFO: Namespace(config_file='', opts=[])
2019-06-27 11:40:13,190 reid_baseline INFO: Running with config:
DATALOADER:
  NUM_INSTANCE: 4
  NUM_WORKERS: 8
  SAMPLER: softmax_triplet
DATASETS:
  NAMES: market1501
  ROOT_DIR:reid-strong-basseline/data
INPUT:
  PADDING: 10
  PIXEL_MEAN: [0.485, 0.456, 0.406]
  PIXEL_STD: [0.229, 0.224, 0.225]
  PROB: 0.5
  RE_PROB: 0.5
  SIZE_TEST: [256, 128]
  SIZE_TRAIN: [256, 128]
MODEL:
  DEVICE: cuda
  DEVICE_ID: 2
  IF_LABELSMOOTH: on
  IF_WITH_CENTER: no
  LAST_STRIDE: 1
  METRIC_LOSS_TYPE: triplet_cluster
  NAME: resnet50
  NECK: bnneck
  PRETRAIN_CHOICE: others
  PRETRAIN_PATH:reid-strong-baseline/checkpoint_mod/train/market1501_1/resnet50_model_120.pth
OUTPUT_DIR: ../checkpoint_mod/train/market1501_triplet_cluster
SOLVER:
  BASE_LR: 0.00035
  BIAS_LR_FACTOR: 1
  CENTER_LOSS_WEIGHT: 0.0005
  CENTER_LR: 0.5
  CHECKPOINT_PERIOD: 40
  CLUSTER_MARGIN: 0.3
  EVAL_PERIOD: 40
  GAMMA: 0.1
  IMS_PER_BATCH: 64
  LOG_PERIOD: 20
  MARGIN: 0.3
  MAX_EPOCHS: 120
  MOMENTUM: 0.9
  OPTIMIZER_NAME: Adam
  RANGE_ALPHA: 0
  RANGE_BETA: 1
  RANGE_K: 2
  RANGE_LOSS_WEIGHT: 1
  RANGE_MARGIN: 0.3
  STEPS: (40, 70)
  WARMUP_FACTOR: 0.1
  WARMUP_ITERS: 10
  WARMUP_METHOD: linear
  WEIGHT_DECAY: 0.0005
  WEIGHT_DECAY_BIAS: 0.0005
TEST:
  FEAT_NORM: yes
  IMS_PER_BATCH: 128
  NECK_FEAT: after
  RE_RANKING: no
  WEIGHT: ...../reid-strong-baseline/checkpoint_mod/train/market1501_1/resnet50_model_120.pth
2019-06-27 15:51:02,642 reid_baseline.train INFO: mAP: 66.7%
2019-06-27 15:51:02,642 reid_baseline.train INFO: CMC curve, Rank-1  :83.6%
2019-06-27 15:51:02,642 reid_baseline.train INFO: CMC curve, Rank-5  :94.2%
2019-06-27 15:51:02,642 reid_baseline.train INFO: CMC curve, Rank-10 :96.4%
I didn't load the .yml file, just all the parameters set in defaults.py, then run in pycharm.
Thank you for your reply.

**Comment by michuanhaohao** [comment]
@Youskrpig What is the number of P and K?

**Comment by michuanhaohao** [comment]
@Simplesss You used 'triplet_cluster'. Cluster loss is usefulness and we forgot delete it. Please try 'triplet' or 'triplet_center' again.

**Comment by Youskrpig** [comment]
_C.SOLVER.IMS_PER_BATCH = 64 （P=16 K=4）
In deafult.py, the parameters are set as follow:

_C.MODEL.DEVICE = "cuda"
_C.MODEL.DEVICE_ID = '4'
_C.MODEL.NAME = 'resnet50'
_C.MODEL.LAST_STRIDE = 1
_C.MODEL.PRETRAIN_CHOICE = 'imagenet'
_C.MODEL.NECK = 'bnneck'
_C.MODEL.IF_WITH_CENTER = 'no'
_C.MODEL.METRIC_LOSS_TYPE = 'triplet'
_C.MODEL.IF_LABELSMOOTH = 'on'
_C.INPUT.SIZE_TRAIN = [384, 128]
_C.INPUT.SIZE_TEST = [384, 128]
_C.INPUT.PROB = 0.5
_C.INPUT.RE_PROB = 0.5
_C.INPUT.PIXEL_MEAN = [0.485, 0.456, 0.406]
_C.INPUT.PIXEL_STD = [0.229, 0.224, 0.225]
_C.INPUT.PADDING = 10
_C.DATASETS.NAMES = ('market1501')
_C.DATALOADER.NUM_WORKERS = 0
_C.DATALOADER.SAMPLER = 'softmax'
_C.DATALOADER.NUM_INSTANCE = 16
_C.SOLVER.OPTIMIZER_NAME = "Adam"
_C.SOLVER.MAX_EPOCHS = 120
_C.SOLVER.BASE_LR = 3.5e-4
_C.SOLVER.BIAS_LR_FACTOR = 2
_C.SOLVER.MOMENTUM = 0.9
_C.SOLVER.MARGIN = 0.3
_C.SOLVER.CLUSTER_MARGIN = 0.3
_C.SOLVER.CENTER_LR = 0.5
_C.SOLVER.CENTER_LOSS_WEIGHT = 0.0005
_C.SOLVER.WEIGHT_DECAY = 0.0005
_C.SOLVER.WEIGHT_DECAY_BIAS = 0.
_C.SOLVER.GAMMA = 0.1
_C.SOLVER.STEPS = (40, 70)

# warm up factor
_C.SOLVER.WARMUP_FACTOR = 1.0
# iterations of warm up
_C.SOLVER.WARMUP_ITERS = 10
# method of warm up, option: 'constant','linear'
_C.SOLVER.WARMUP_METHOD = "linear"

About warmup, My environment is python2, So i slightly changed the code to ensure the learning rate  as the paper said.

# epoch number of saving checkpoints
_C.SOLVER.CHECKPOINT_PERIOD = 120
_C.SOLVER.LOG_PERIOD = 100
_C.SOLVER.EVAL_PERIOD = 10

_C.SOLVER.IMS_PER_BATCH = 64
_C.TEST.IMS_PER_BATCH = 128
_C.TEST.RE_RANKING = 'no'
_C.TEST.NECK_FEAT = 'after'
_C.TEST.FEAT_NORM = 'yes'

  Is there any possibility because my environment is python 2 ?

**Comment by Youskrpig** [comment]
_C.SOLVER.IMS_PER_BATCH = 64 （P=16 K=4）
In deafult.py, the parameters are set as follow:

_C.MODEL.DEVICE = "cuda"
_C.MODEL.DEVICE_ID = '4'
_C.MODEL.NAME = 'resnet50'
_C.MODEL.LAST_STRIDE = 1
_C.MODEL.PRETRAIN_CHOICE = 'imagenet'
_C.MODEL.NECK = 'bnneck'
_C.MODEL.IF_WITH_CENTER = 'no'
_C.MODEL.METRIC_LOSS_TYPE = 'triplet'
_C.MODEL.IF_LABELSMOOTH = 'on'
_C.INPUT.SIZE_TRAIN = [384, 128]
_C.INPUT.SIZE_TEST = [384, 128]
_C.INPUT.PROB = 0.5
_C.INPUT.RE_PROB = 0.5
_C.INPUT.PIXEL_MEAN = [0.485, 0.456, 0.406]
_C.INPUT.PIXEL_STD = [0.229, 0.224, 0.225]
_C.INPUT.PADDING = 10
_C.DATASETS.NAMES = ('market1501')
_C.DATALOADER.NUM_WORKERS = 0
_C.DATALOADER.SAMPLER = 'softmax'
_C.DATALOADER.NUM_INSTANCE = 16
_C.SOLVER.OPTIMIZER_NAME = "Adam"
_C.SOLVER.MAX_EPOCHS = 120
_C.SOLVER.BASE_LR = 3.5e-4
_C.SOLVER.BIAS_LR_FACTOR = 2
_C.SOLVER.MOMENTUM = 0.9
_C.SOLVER.MARGIN = 0.3
_C.SOLVER.CLUSTER_MARGIN = 0.3
_C.SOLVER.CENTER_LR = 0.5
_C.SOLVER.CENTER_LOSS_WEIGHT = 0.0005
_C.SOLVER.WEIGHT_DECAY = 0.0005
_C.SOLVER.WEIGHT_DECAY_BIAS = 0.
_C.SOLVER.GAMMA = 0.1
_C.SOLVER.STEPS = (40, 70)

_C.SOLVER.WARMUP_FACTOR = 1.0
_C.SOLVER.WARMUP_ITERS = 10
_C.SOLVER.WARMUP_METHOD = "linear"

About warmup, My environment is python2, So i slightly changed the code to ensure the learning rate as the paper said.

_C.SOLVER.IMS_PER_BATCH = 64
_C.TEST.IMS_PER_BATCH = 128
_C.TEST.RE_RANKING = 'no'
_C.TEST.NECK_FEAT = 'after'
_C.TEST.FEAT_NORM = 'yes'

Is there any possibility because my environment is python 2 ?

**Comment by michuanhaohao** [comment]
@Youskrpig Maybe. I remember that the number of iterations is 183 or 186, not 203. Please check the dataset.

**Comment by Youskrpig** [comment]
Ummm, the number of Training image of Market 1501 is 12936, the batchsize is 64. So the iteration equals 12936/64 ≈ 202.  Ok i will check

**Comment by michuanhaohao** [comment]
@Youskrpig Some images are dropped because of the long-tailed distribution. It needs 16 identities (at least 4 images per ID) to get a batch.

**Comment by shaoniangu** [comment]
@Youskrpig Hi, I think you should check your configurations with our .yml file carefully if you want to reproduce the same results with our paper. I found there are some different configurations between the settings your comments provide and our settings merged with softmax_triplet_with_center.yml file, such as 
(1)DATALOADER.SAMPLER. Ours is 'softmax_triplet', yours is 'softmax'. This parameter determines the sample method. Sampler of 'softmax' method samples data in order from data set, while sampler of 'softmax_triplet' method samples data using P×K manner.
(2)DATALOADER.NUM_INSTANCE. This parameter determines the value of K. Ours is 4, yours is 16.
(3)I do not see your settings about BNNeck, such as which feature (before or after) BNNeck is choosed for train, which may influence a lot on the result.
The parameters our experiments uses are result of the merging of defaults.py, the .yml file and the command line. You should check carefully. 
Thank you for your issue.

**Comment by Youskrpig** [comment]
@michuanhaohao @shaoniangu  Thank you very much O(∩_∩)O
